### PR TITLE
welcome-screen.sh: prevent login hang when a network mount is unreachable

### PR DESCRIPTION
**Pull request by damon-downing**
### Problem

`welcome-screen.sh` calls `screenfetch`/`neofetch` on every login to display the system-info banner. Both tools' disk-usage check runs `df --total`, which touches every mounted filesystem with no exclusions and no timeout.

If a Proxmox NFS or CIFS storage's server goes down while the Proxmox host itself stays up, that `df` call can block indefinitely inside the kernel (uninterruptible sleep, D-state). Because the script runs during MOTD generation, this can block SSH, the Proxmox web shell, and console logins.

### Fix

Do not perform disk discovery in the login/MOTD path at all.

`screenfetch` and `neofetch` normally call `df` to calculate disk usage. On hosts with NFS/CIFS mounts, that can block indefinitely when a remote filesystem is unavailable.

The welcome screen now:

- disables the disk field for both `screenfetch` and `neofetch`
- wraps each command with a bounded 10-second timeout as an additional safeguard
- keeps the correct `bash` shell value for `screenfetch`
- performs no NFS/CIFS reachability probes or `/proc/mounts` parsing

This avoids touching remote filesystems during login entirely. The existing local version-cache and configuration checks remain unchanged.

### Tested

- Login completes normally with healthy local storage.
- Login remains bounded when remote NFS/CIFS storage is unavailable.
- `screenfetch` and `neofetch` no longer perform disk discovery.
- The `screenfetch` shell field remains correct.
- The remaining welcome-banner fields are unaffected.
- `bash -n`, targeted ShellCheck for `welcome-screen.sh`, and `git diff --check` pass.

**-Damon1974**

### CODE REVIEW

**Comment by damon-downing** [comment]
Note: the ShellCheck failure here is pre-existing lint debt unrelated to this change — every recent push to develop fails the same check (e.g. #1155, #1154, #1153...). None of the flagged lines are in welcome-screen.sh.

**-Damon1974**

**Comment by BassT23** [comment]
Thanks for the PR — we reviewed #318 against the current `develop` branch and confirmed the underlying issue and your general approach are valid.

We definitely want to keep your contribution as the basis for the fix rather than replacing it with a separate implementation.

During review we found one way to make the login path a bit more robust: instead of probing NFS/CIFS servers first and only disabling disk detection when a server is unreachable, we would prefer to disable the disk check from `screenfetch` / `neofetch` unconditionally in the MOTD path.

The reason is that a server can still accept TCP connections on 2049/445 while the actual mount or filesystem access is stuck. In that case `df` can still block in D-state, and `timeout` cannot reliably recover once that happens.

We prepared and tested a small follow-up change locally:

* remove the NFS/CIFS reachability probes
* always run `screenfetch` with disk detection disabled
* run `neofetch` with `--disable disk`
* keep the `shell_type="bash"` fix
* keep the local version-cache and config checks unchanged
* no `df`, `/proc/mounts`, or remote-storage access remains in the login path

Tests passed for `screenfetch`, `neofetch`, neither tool installed, `bash -n`, ShellCheck on `welcome-screen.sh`, and `git diff --check`.

At the moment we cannot push the follow-up directly because the PR has `maintainerCanModify=false`.

If you are happy with this direction, could you either enable maintainer edits for the PR or apply the equivalent change on your branch? We’re happy to continue reviewing and improving this together before anything is merged.

Thanks again for finding and documenting the issue so thoroughly.

**Comment by BassT23** [comment]
Small clarification to my previous comment:

When I say **“we”**, I mean **me together with my AI-assisted development/review team**. I use AI tools extensively for code review, testing and preparing changes, but I make the project decisions and review the results before anything is accepted.

Also, if you enable **maintainer edits** for this PR, we can add the already prepared and tested follow-up commit directly to your existing PR branch. That would keep your original commit and authorship intact while allowing us to improve #318 together instead of creating a separate implementation.

**Comment by damon-downing** [comment]
Thanks for digging into this so thoroughly, and glad we're aligned on the direction. I've had some time to sit with it, and honestly I think you're right. 

This is fundamentally an updater tool, not a monitoring tool. There's no threshold-based disk alerting anywhere in the codebase (the only other `df `usage, in `update.sh`, just gates available space during an update/snapshot, it's not user-facing). 

Given that, the `Disk:` line in the banner was always cosmetic, so trading it away entirely to fully close the D-state risk is the right call. I appreciate you catching that edge case, the "TCP-connect-but-NFS-hung" scenario isn't something I was able to test for.

One hiccup on my end:  `downing-labs` is an org-owned GitHub account, and GitHub doesn't offer "allow edits by maintainers" for PRs from org-owned forks, only individual-accounts. So the checkbox genuinely isn't there for me to check.

If you paste the diff (or `git format-patch` output) here in a comment, I'll grab it and push it to my branch myself, happy to just apply what you've already tested.

Thanks again for taking the time to review this properly.

**-Damon1974**

**Comment by BassT23** [comment]
Thanks for the review and for the original fix in this PR. I prepared a maintainer follow-up based on the current PR HEAD.

The follow-up is exactly one commit and changes only `welcome-screen.sh`. It:

- removes the NFS/CIFS TCP reachability probes and `/proc/mounts` parsing
- disables disk discovery for `screenfetch` and `neofetch`
- preserves the existing `shell_type="bash"`, local version-cache, and `READ_WRITE_CONFIG` logic

The patch was applied locally with `git am` on the current PR HEAD and the targeted tests are green. To apply it:

```sh
git checkout fix/welcome-screen-disk-hang
git am pr-318-followup.patch
```

If application needs to be aborted:

```sh
git am --abort
```

Full format-patch:

```diff
From f75174c0ecbc6872cda2e2ea30abe3d4b456ee8a Mon Sep 17 00:00:00 2001
From: BassT <github@familiebecke.de>
Date: Sat, 22 Aug 2026 11:22:06 +0200
Subject: [PATCH] maint: avoid remote disk discovery in welcome screen

---
 welcome-screen.sh | 62 ++++++++---------------------------------------
 1 file changed, 10 insertions(+), 52 deletions(-)

diff --git a/welcome-screen.sh b/welcome-screen.sh
index b0859a5..9cbaa06 100644
--- a/welcome-screen.sh
+++ b/welcome-screen.sh
@@ -23,35 +23,6 @@ OR="\e[1;33m"
 GN="\e[1;92m"
 CL="\e[0m"
 
-# NFS/CIFS reachability check: avoids calling df on an unreachable network
-# mount, which can hang indefinitely (D-state) and block MOTD/login host-wide.
-CHECK_NFS_REACHABILITY () {
-  DOWN_MOUNTS=()
-  local spec mnt fstype server port
-
-  while read -r spec mnt fstype _; do
-    case "$fstype" in
-      nfs|nfs4)
-        server="${spec%%:*}"
-        port=2049
-        ;;
-      cifs|smbfs|smb3)
-        server="${spec#\\\\}"
-        server="${server%%\\*}"
-        server="${server#//}"
-        server="${server%%/*}"
-        port=445
-        ;;
-      *)
-        continue
-        ;;
-    esac
-    if ! timeout 2 bash -c "exec 3<>/dev/tcp/${server}/${port}" 2>/dev/null; then
-      DOWN_MOUNTS+=("$mnt")
-    fi
-  done < <(awk '{print $1, $2, $3}' /proc/mounts)
-}
-
 # shellcheck disable=SC1091
 . "$LOCAL_FILES/tag-filter.sh"
 
@@ -142,30 +113,17 @@ TIME_CALCULTION () {
   MINUTES=$(( (NOW - MOD) / 60 ))
 }
 
-# Welcome
-CHECK_NFS_REACHABILITY
-if [[ ${#DOWN_MOUNTS[@]} -eq 0 ]]; then
-  if [[ -f /usr/bin/screenfetch ]]; then
-    echo && timeout 10 screenfetch -o 'shell_type="bash"' && echo
-  elif [[ -f /usr/bin/neofetch ]]; then
-    echo
-    timeout 10 neofetch
-  else
-    echo
-  fi
+# Welcome. Disk discovery is deliberately disabled: screenfetch/neofetch
+# otherwise call df across every mounted filesystem, and a dead NFS/CIFS
+# server can leave that syscall in uninterruptible kernel sleep. A login/MOTD
+# path must not probe remote storage or rely on a network reachability guess.
+if [[ -f /usr/bin/screenfetch ]]; then
+  echo && timeout 10 screenfetch -d '-disk' -o 'shell_type="bash"' && echo
+elif [[ -f /usr/bin/neofetch ]]; then
+  echo
+  timeout 10 neofetch --disable disk
 else
-  if [[ -f /usr/bin/screenfetch ]]; then
-    echo && timeout 10 screenfetch -d '-disk' -o 'shell_type="bash"' && echo
-  elif [[ -f /usr/bin/neofetch ]]; then
-    echo
-    timeout 10 neofetch --disable disk
-  else
-    echo
-  fi
-  echo -e "${OR}*** One or more disks are offline - can't check disk usage ***${CL}"
-  for m in "${DOWN_MOUNTS[@]}"; do
-    echo -e "${OR}       $m: OFFLINE${CL}"
-  done
+  echo
 fi
 VERSION_CHECK
 READ_WRITE_CONFIG
-- 
2.43.0

```

**Comment by damon-downing** [comment]
@BassT23 

Applied, thanks. `git diff --stat` came back as `10 insertions(+), 52 deletions(-)`, an exact match to what you reported, so the result should be identical to what you tested locally.

Pushed as `3d51a8d` on top of the original commit.

**-Damon1974**
